static/js/containers/pages/CheckoutPage_test.js

```diff
@@ -102,6 +104,132 @@ describe("CheckoutPage", () => {
+        {
+          location: {
+            search: `product=${productId}`
+          }
+        }
+      )
+
+      sinon.assert.calledWith(
+        helper.handleRequestStub,
+        "/api/basket/",
+        "PATCH",
+        {
+          body:        { items: [{ id: productId }] },
+          credentials: undefined,
+          headers:     {
+            "X-CSRFTOKEN": null
+          }
+        }
+      )
+      assert.equal(inner.state().errors, hasError ? "error" : null)
+    })
+  })
+
+  //
+  ;[[true, false], [true, true], [false, false], [false, true]].forEach(
+    ([hasValidProductId, hasValidCoupon]) => {
+      it(`updates the basket with a ${
+        hasValidProductId ? "" : "in"
+      }valid product id and a ${
+        hasValidCoupon ? "" : "in"
+      }valid coupon code from the query parameter`, async () => {
+        const couponCode = "codeforcoupon"
+        const productId = 12345
+        const couponPayload = {
+          body:        { coupons: [{ code: couponCode }] },
+          credentials: undefined,
+          headers:     {
+            "X-CSRFTOKEN": null
+          }
+        }
+        const productPayload = {
+          body:        { items: [{ id: productId }] },
+          credentials: undefined,
+          headers:     {
+            "X-CSRFTOKEN": null
+          }
+        }
+
+        if (!hasValidProductId) {
+          helper.handleRequestStub
+            .withArgs("/api/basket/", "PATCH", productPayload)
+            .returns({
+              status: 400,
+              body:   {
+                errors: "product error"
+              }
+            })
+        }
+        if (!hasValidCoupon) {
+          helper.handleRequestStub
+            .withArgs("/api/basket/", "PATCH", couponPayload)
+            .returns({
+              status: 400,
+              body:   {
+                errors: "coupon error"
+              }
+            })
+        }
+        const { inner } = await renderPage(
+          {},
+          {
+            location: {
+              search: `product=${productId}&code=${couponCode}`
+            }
+          }
+        )
+        // wait for componentDidMount to resolve
+        await wait(15)
+        sinon.assert.calledWith(
+          helper.handleRequestStub,
+          "/api/basket/",
+          "PATCH",
+          productPayload
+        )
+        if (hasValidProductId) {
+          sinon.assert.calledWith(
+            helper.handleRequestStub,
+            "/api/basket/",
+            "PATCH",
+            couponPayload
+          )
+        } else {
+          sinon.assert.neverCalledWith(
+            helper.handleRequestStub,
+            "/api/basket/",
+            "PATCH",
+            couponPayload
+          )
+        }
+
+        assert.equal(
+          inner.state().errors,
+          !hasValidProductId
+            ? "product error"
+            : !hasValidCoupon
+              ? "coupon error"
+              : null
+        )
+      })
+    }
```

Comment:
One comment off the top: I would rather have multiple test cases with some repeated code than a parametrized test case with this many lines and this many logical paths to reason about

Reply:
...actually I changed my mind. I think this can be cleaned in a way that makes the parametrization a little more readable and maintainable

```javascript
describe("mount", () => {
    const productError = "product error",
      couponError = "coupon error",
      couponCode = "codeforcoupon",
      productId = 12345
    let couponPayload,productPayload

    beforeEach(() => {
      couponPayload = {
        body: {coupons: [{code: couponCode}]},
        credentials: undefined,
        headers: {
          "X-CSRFTOKEN": null
        }
      }
      productPayload = {
        body: {items: [{id: productId}]},
        credentials: undefined,
        headers: {
          "X-CSRFTOKEN": null
        }
      }
    })

    //
    ;[
      [true, false, couponError],
      [true, true, null],
      [false, false, productError],
      [false, true, productError]
    ].forEach(
      ([hasValidProductId, hasValidCoupon, expError]) => {
        it(`...`, async () => {
          if (!hasValidProductId) {
            helper.handleRequestStub
              //...
              .returns({
                //...
                body: {
                  errors: productError
                }
              })
          }
          if (!hasValidCoupon) {
            helper.handleRequestStub
              //...
              .returns({
                //...
                body: {
                  errors: couponError
                }
              })
          }
          //...

          assert.equal(
            helper.handleRequestStub.calledWith(
              "/api/basket/",
              "PATCH",
              couponPayload
            ),
            hasValidProductId
          )

          assert.equal(
            inner.state().errors,
            expError
          )
        })
      }
    )
  })
```